src/server/frontend/wayland/wayland_connector.cpp

```diff
@@ -1954,6 +1961,532 @@ class WlShell : public wayland::Shell
+    {
+    }
+
+    void grab(struct wl_resource* seat, uint32_t serial) override
+    {
+        (void)seat, (void)serial;
+        // TODO
+    }
+
+    void destroy() override
+    {
+        wl_resource_destroy(resource);
+    }
+};
+
+class XdgSurfaceV6EventSink : public BasicSurfaceEventSink
+{
+public:
+    using BasicSurfaceEventSink::BasicSurfaceEventSink;
+
+    XdgSurfaceV6EventSink(WlSeat* seat, wl_client* client, wl_resource* target, wl_resource* event_sink,
+        std::shared_ptr<bool> const& destroyed) :
+        BasicSurfaceEventSink(seat, client, target, event_sink),
+        destroyed{destroyed}
+    {
+        auto const serial = wl_display_next_serial(wl_client_get_display(client));
+        post_configure(serial);
+    }
+
+    void send_resize(geometry::Size const& new_size) const override
+    {
+        if (window_size != new_size)
+        {
+            auto const serial = wl_display_next_serial(wl_client_get_display(client));
+            notify_resize(new_size);
+            post_configure(serial);
+        }
+    }
+
+    std::function<void(geometry::Size const& new_size)> notify_resize = [](auto){};
+
+private:
+    void post_configure(int serial) const
+    {
+        seat->spawn(run_unless(destroyed, [event_sink= event_sink, serial]()
+            {
+                wl_resource_post_event(event_sink, 0, serial);
+            }));
+    }
+
+    std::shared_ptr<bool> const destroyed;
+};
+
+struct XdgSurfaceV6 : wayland::XdgSurfaceV6, WlAbstractMirWindow
+{
+    XdgSurfaceV6* get_xdgsurface(wl_resource* surface) const
+    {
+        auto* tmp = wl_resource_get_user_data(surface);
+        return static_cast<XdgSurfaceV6*>(static_cast<wayland::XdgSurfaceV6*>(tmp));
+    }
+
+    XdgSurfaceV6(wl_client* client,
+        wl_resource* parent,
+        uint32_t id,
+        wl_resource* surface,
+        std::shared_ptr<mf::Shell> const& shell,
+        WlSeat& seat) :
+        wayland::XdgSurfaceV6(client, parent, id),
+        WlAbstractMirWindow{client, surface, resource, shell},
+        parent{parent},
+        shell{shell},
+        sink{std::make_shared<XdgSurfaceV6EventSink>(&seat, client, surface, resource, destroyed)}
+    {
+        WlAbstractMirWindow::sink = sink;
+    }
+
+    ~XdgSurfaceV6() override
+    {
+        auto* const mir_surface = get_wlsurface(surface);
+        mir_surface->set_role(&nullwlmirwindow);
+    }
+
+    void destroy() override
+    {
+        wl_resource_destroy(resource);
+    }
+
+    void get_toplevel(uint32_t id) override
+    {
+        new XdgToplevelV6{client, parent, id, shell, this};
+        auto* const mir_surface = get_wlsurface(surface);
+        mir_surface->set_role(this);
+    }
+
+    void get_popup(uint32_t id, struct wl_resource* parent, struct wl_resource* positioner) override
+    {
+        auto* tmp = wl_resource_get_user_data(positioner);
+        auto const* const pos =  static_cast<XdgPositionerV6*>(static_cast<wayland::XdgPositionerV6*>(tmp));
```

Comment:
Why the 2 casts? The outer one alone should be enough

Reply:
I'm sure it is always optimized out, but the standardise doesn't support your assertion.

Reply:
I did check with gcc, it was happy just using the outer. No objection though

Reply:
You can static_cast from "void*" to "anything*" without the compiler complaining - you're telling it you know what you're doing.

A cast from Base* to Derived* isn't guaranteed to be a no-op (it happens to be a no-op for _single_, _non-virtual inheritance_ on every implementation I've used but I don't believe the standard requires it).

Reply:
Sure. I'm aware it should be a no-op. I'm just more considering the code readability. Again, no objection